b2/src/main/java/org/jclouds/b2/domain/B2Object.java

```diff
@@ -0,0 +1,62 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.b2.domain;
+
+import java.util.Date;
+import java.util.Map;
+
+import org.jclouds.io.Payload;
+import org.jclouds.javax.annotation.Nullable;
+import org.jclouds.json.SerializedNames;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+import com.google.common.hash.Hashing;
+
+@AutoValue
+public abstract class B2Object {
+   /** File ID, only set by the server. */
+   @Nullable public abstract String fileId();
+   public abstract String fileName();
+   public abstract String contentSha1();
+   public abstract Map<String, String> fileInfo();
+   @Nullable public abstract Payload payload();
+   /** Upload timestamp, only set by the server. */
+   @Nullable public abstract Date uploadTimestamp();
+
+   @Nullable public abstract Action action();
+   @Nullable public abstract String accountId();
+   @Nullable public abstract String bucketId();
+   @Nullable public abstract Long contentLength();
+   @Nullable public abstract String contentType();
+
+   public static B2Object create(String fileName, String contentSha1, Map<String, String> fileInfo, @Nullable Payload payload) {
+      return createResponse(null, fileName, contentSha1, fileInfo, payload, null);
+   }
+
+   /** Create a response to a request.  Most external callers should call create. */
+   public static B2Object createResponse(@Nullable String fileId, String fileName, String contentSha1, Map<String, String> fileInfo, @Nullable Payload payload, @Nullable Date uploadTimestamp) {
+      Preconditions.checkArgument(contentSha1.length() == Hashing.sha1().bits() / 4,
+            "Expected hash length of %s, was: %s", Hashing.sha1().bits() / 4, contentSha1.length());
+      return new AutoValue_B2Object(fileId, fileName, contentSha1, fileInfo, payload, uploadTimestamp, null, null, null, null, null);
+   }
+
+   @SerializedNames({"fileId", "fileName", "accountId", "bucketId", "contentLength", "contentSha1", "contentType", "fileInfo", "action", "uploadTimestamp", "payload"})
+   public static B2Object createHeadResponse(String fileId, String fileName, String accountId, String bucketId, long contentLength, String contentSha1, String contentType, Map<String, String> fileInfo, Action action, long uploadTimestamp, @Nullable Payload payload) {
+      return new AutoValue_B2Object(fileId, fileName, contentSha1, fileInfo, null, new Date(uploadTimestamp), action, accountId, bucketId, contentLength, contentType);
+   }
```

Comment:
In all factory methods, enforce immutable collections. Use ImmutableMap.copyOf for fileInfo.